File: src/main/java/io/github/mdsimmo/bomberman/game/GamePlayer.java
package io.github.mdsimmo.bomberman.game;

import io.github.mdsimmo.bomberman.Bomberman;
import io.github.mdsimmo.bomberman.events.*;
import io.github.mdsimmo.bomberman.messaging.Formattable;
import io.github.mdsimmo.bomberman.messaging.Message;
import io.github.mdsimmo.bomberman.messaging.Text;
import org.bukkit.Bukkit;
import org.bukkit.GameMode;
import org.bukkit.Location;
import org.bukkit.Server;
import org.bukkit.attribute.Attribute;
import org.bukkit.block.Block;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.event.*;
import org.bukkit.event.block.Action;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.EntityRegainHealthEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerMoveEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.player.PlayerRespawnEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

import javax.annotation.Nonnull;
import java.util.List;
import java.util.stream.Collectors;

public class GamePlayer implements Formattable, Listener {

    private static JavaPlugin plugin = Bomberman.instance;

    public static void spawnGamePlayer(@Nonnull Player player, @Nonnull  Game game, @Nonnull Location start) {

        GamePlayer gamePlayer = new GamePlayer(player, game);

        player.getServer().getPluginManager().registerEvents(gamePlayer, plugin);

        // Initialise the player for the game
        player.teleport(start.clone().add(0.5, 0.5, 0.5));
        player.setGameMode(GameMode.SURVIVAL);
        player.setHealth(game.getSettings().getLives());
        player.getAttribute(Attribute.GENERIC_MAX_HEALTH).setBaseValue(game.getSettings().getLives());
        // if setHealthScale is not delayed, it can sometimes cause the client side to think they died?!?!?
        Bukkit.getScheduler().scheduleSyncDelayedTask(plugin, () ->player.setHealthScale(game.getSettings().getLives() * 2));
        player.setExhaustion(0);
        player.setFoodLevel(100000); // just a big number
        player.getInventory().clear();
        for (ItemStack stack : game.getSettings().getInitialItems()) {
            ItemStack s = stack.clone();
            player.getInventory().addItem(s);
        }

        gamePlayer.removePotionEffects();
        player.addScoreboardTag("bm_player");
    }

    private final Player player;
    private final Game game;

    private boolean immunity = false;

    private final ItemStack[] spawnInventory;
    private final Location spawn;
    private final int spawnHunger;
    private final GameMode spawnGameMode;
    private final double spawnHealth;
    private final double spawnMaxHealth;
    private final double spawnHealthScale;

    private GamePlayer(Player player, Game game) {
        this.player = player;
        this.game = game;

        // remember the player stats
        spawnHealth = player.getHealth();
        spawnGameMode = player.getGameMode();
        spawnHealthScale = player.getHealthScale();
        spawnMaxHealth = player.getAttribute(Attribute.GENERIC_MAX_HEALTH).getBaseValue();
        spawn = player.getLocation();
        spawnHunger = player.getFoodLevel();
        spawnInventory = player.getInventory().getContents();
    }

    /**
     * Removes the player from the game and removes any hooks to this player. Treats the player like they disconnected
     * from the server.
     */
    private void resetStuffAndUnregister() {
        reset();
        HandlerList.unregisterAll(this);
    }

    private void reset() {
        // remove items in the direct vicinity (prevents player dropping items at spawn)
        player.getWorld().getNearbyEntities(player.getLocation(), 1, 2, 1).stream()
                .filter(it -> it instanceof ItemStack)
                .forEach(Entity::remove);

        player.getAttribute(Attribute.GENERIC_MAX_HEALTH).setBaseValue(spawnMaxHealth);
        player.setHealthScale(spawnHealthScale);
        player.setHealth(spawnHealth);
        player.teleport(spawn);
        player.setGameMode(spawnGameMode);
        player.getInventory().setContents(spawnInventory);
        player.setFoodLevel(spawnHunger);

        player.removeScoreboardTag("bm_player");

        removePotionEffects();
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    public void onPlayerLeaveGameEvent(BmPlayerLeaveGameIntent e) {
        if (e.getPlayer() != player)
            return;
        if (player.isDead()) {
            // Give player their stuff back when they respawn
            // Attempting to do this when player is dead causes very strange bugs
            Bukkit.getPluginManager().registerEvents(new Listener() {
                @EventHandler
                public void onPlayerRespawn(PlayerRespawnEvent e) {
                    e.setRespawnLocation(spawn);
                    reset();
                    HandlerList.unregisterAll(this);
                }
            }, plugin);
            HandlerList.unregisterAll(this);
        } else {
            resetStuffAndUnregister();
        }
        e.setHandled();
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    public void onGameTerminated(BmGameTerminatedIntent e) {
        if (e.getGame() != this.game)
            return;
        resetStuffAndUnregister();
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    public void onPlayerMoved(PlayerMoveEvent e) {
        if (e.getPlayer() != player)
            return;
        BmPlayerMovedEvent bmEvent = new BmPlayerMovedEvent(game, player, e.getFrom(), e.getTo());
        Bukkit.getPluginManager().callEvent(bmEvent);
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    public void onExplosion(BmExplosionEvent e) {
        // TODO duplicate code: both GamePlayer and Explosion do touching checks
        if (Explosion.isTouching(player, e.getIgniting().stream().map(b->b.block).collect(Collectors.toSet()))) {
            BmPlayerHitIntent.hit(player, e.getCause());
        }
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    public void onPlayerHit(BmPlayerHitIntent e) {
        if (e.getPlayer() != this.player)
            return;

        BmPlayerHurtIntent.run(game, player, e.getCause());
        e.setHandled();
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    public void onPlayerHurtWithImmunity(BmPlayerHurtIntent e) {
        if (e.getPlayer() != player)
            return;

        if (immunity)
            e.setCancelled(true);
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    public void onPlayerDamaged(BmPlayerHurtIntent e) {
        if (e.getPlayer() != player)
            return;

        if (player.getHealth() > 1) {
            player.damage(1);
            immunity = true;
            player.setFireTicks(22);
            Bukkit.getScheduler().scheduleSyncDelayedTask(plugin, () -> {
                immunity = false;
                player.setFireTicks(0);
                // Call the player move event to recheck damage required
                Bukkit.getPluginManager().callEvent(new BmPlayerMovedEvent(game, player, player.getLocation(), player.getLocation()));
            }, 22); // 22 is slightly longer than 20 ticks a bomb is active for
        } else {
            BmPlayerKilledIntent.kill(game, player, e.getAttacker());
        }
        e.setHandled();
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    public void onPlayerKilledInGame(BmPlayerKilledIntent e) {
        if (e.getPlayer() != player)
            return;
        player.setHealth(0);
        BmPlayerLeaveGameIntent.leave(player);
        e.setHandled();
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
    public void onPlayerPlaceBlock(BlockPlaceEvent e) {
        if (e.getPlayer() != player)
            return;
        Block b = e.getBlock();
        // create a bomb when placing tnt
        if (b.getType() == game.getSettings().getBombItem()) {
            if (!Bomb.spawnBomb(game, player, b)) {
                e.setCancelled(true);
            }
        }
    }

    @EventHandler(priority = EventPriority.HIGH)
    public void onPlayerLogout(PlayerQuitEvent e) {
        if (e.getPlayer() == player) {
            BmPlayerLeaveGameIntent.leave(player);
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onPlayerRegen(EntityRegainHealthEvent e) {
        if (e.getEntity() == player) {
            if (e.getRegainReason() == EntityRegainHealthEvent.RegainReason.MAGIC) {
                e.setAmount(1);
            } else {
                e.setCancelled(true);
            }
        }
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOWEST)
    public void onPlayerDamaged(EntityDamageEvent e) {
        if (e.getEntity() != player)
            return;
        // Allow custom damage events (ie. from plugin)
        if (e.getCause() == EntityDamageEvent.DamageCause.CUSTOM)
            return;
        // Player cannot be burnt or hurt during game play
        e.setCancelled(true);
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    public void onPlayerJoinGame(BmPlayerJoinGameIntent e) {
        // Cannot join two games at once
        if (e.getPlayer() == player) {
            e.cancelFor(Text.JOIN_ALREADY_JOINED
                    .with("game", e.getGame())
                    .with("player", player)
                    .format());
        }
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    public void onGameStopped(BmRunStoppedIntent e) {
        BmPlayerLeaveGameIntent.leave(player);
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    public void onPlayerBreakBlockWithWrongTool(PlayerInteractEvent e) {
        // Disable player from breaking a block
        if (e.getPlayer() != player)
            return;
        if (e.getAction() != Action.LEFT_CLICK_BLOCK || !e.hasBlock()) {
            // Only care about block breaking events
            return;
        }

        // TODO only let player break block if they have used the correct tool
        // Maybe use CanDestroy tag? - That requires NBT though...

        // Cannot break things with hand
        e.setCancelled(true);
        e.setUseInteractedBlock(Event.Result.DENY);
        e.setUseItemInHand(Event.Result.DENY);
        // apply mining fatigue so player doesn't see block breaking
        e.getPlayer().addPotionEffect(new PotionEffect(PotionEffectType.SLOW_DIGGING, 20, 1));
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    public void onPlayerWon(BmPlayerWonEvent e) {
        if (e.getPlayer() != player)
            return;
        Text.PLAYER_WON.with("player", player).sendTo(player);
        // Let player walk around like a boss
        immunity = true;
    }

    public static int bombStrength(Game game, Player player) {
        int strength = 0;
        for (ItemStack stack : player.getInventory().getContents()) {
            if (stack != null && stack.getType() == game.getSettings().getPowerItem()) {
                strength += stack.getAmount();
            }
        }
        return Math.max(strength, 1);
    }

    private int bombStrength() {
        return bombStrength(game, player);
    }

    private int bombAmount() {
        int strength = 0;
        for (ItemStack stack : player.getInventory().getContents()) {
            if (stack != null && stack.getType() == game.getSettings().getBombItem()) {
                strength += stack.getAmount();
            }
        }
        return Math.max(strength, 1);
    }

    private void removePotionEffects() {
        Server server = player.getServer();
        if (plugin.isEnabled())
            server.getScheduler()
                    .scheduleSyncDelayedTask(plugin, () -> {
                        player.setFireTicks(0);
                        for (PotionEffect effect : player.getActivePotionEffects()) {
                            player.removePotionEffect(effect.getType());
                        }
                    });
    }

    @Override
    public Message format(@Nonnull List<Message> args) {
        if (args.size() == 0)
            return Message.of(player.getName());
        if (args.size() != 1)
            throw new RuntimeException("Players can have at most one argument");
        switch (args.get(0).toString()) {
            case "name":
                return Message.of(player.getName());
            case "lives":
                return Message.of((int) player.getHealth());
            case "power":
                return Message.of(bombStrength());
            case "bombs":
                return Message.of(bombAmount());
            default:
                return null;
        }
    }
}

File: src/main/java/io/github/mdsimmo/bomberman/game/GamePlayer.kt
package io.github.mdsimmo.bomberman.game

import com.sk89q.jnbt.StringTag
import com.sk89q.worldedit.bukkit.BukkitAdapter
import io.github.mdsimmo.bomberman.Bomberman
import io.github.mdsimmo.bomberman.events.*
import io.github.mdsimmo.bomberman.messaging.Text
import org.bukkit.Bukkit
import org.bukkit.GameMode
import org.bukkit.Location
import org.bukkit.attribute.Attribute
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Item
import org.bukkit.entity.Player
import org.bukkit.event.*
import org.bukkit.event.block.Action
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.entity.EntityDamageEvent
import org.bukkit.event.entity.EntityRegainHealthEvent
import org.bukkit.event.player.*
import org.bukkit.inventory.ItemStack
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import java.io.File
import java.util.logging.Level

class GamePlayer private constructor(private val player: Player, private val game: Game) : Listener {

    companion object {
        private val plugin: Bomberman = Bomberman.instance

        @JvmStatic
        fun spawnGamePlayer(player: Player, game: Game, start: Location) {
            // Create the player to store data
            val gamePlayer = GamePlayer(player, game)
            plugin.server.pluginManager.registerEvents(gamePlayer, plugin)

            // Record the player stats in file. Use a file so that server can crash
            val dataFile = YamlConfiguration()
            dataFile["location"] = player.location
            dataFile["gamemode"] = player.gameMode.name.lowercase()
            dataFile["health"] = player.health
            dataFile["health-scale"] = player.healthScale
            dataFile["health-max"] = player.getAttribute(Attribute.GENERIC_MAX_HEALTH)!!.baseValue
            dataFile["food-level"] = player.foodLevel
            dataFile["inventory"] = player.inventory.contents.toList()
            dataFile["is-flying"] = player.isFlying
            dataFile.save(tempDataFile(player))

            // Add a permission group for WorldGuard compatibility (must be done before teleporting)
            try {
                player.addAttachment(plugin, "group.bomberman", true)
            } catch (e: Exception) {
                plugin.logger.log(Level.WARNING, "Unable to add permissions", e)
            }

            // Clear the landing area of items (there should never be any, but we don't want no cheating)
            start.world!!.getNearbyEntities(start, 2.0, 3.0, 2.0)
                    .filterIsInstance<Item>()
                    .forEach{ it.remove() }

            // Initialise the player for the game
            val maxHealth = player.getAttribute(Attribute.GENERIC_MAX_HEALTH)!!
            maxHealth.baseValue = game.settings.lives.toDouble()
            maxHealth.modifiers.forEach { maxHealth.removeModifier(it) }
            player.health = game.settings.lives.toDouble()
            Bukkit.getScheduler().scheduleSyncDelayedTask(plugin) {
                // delayed because it seems to reduce client side death screen glitch
                player.healthScale = game.settings.lives * 2.toDouble()
            }
            if (!player.teleport(start.clone().add(0.5, 0.01, 0.5))) {
                // Teleport failed, revert player state in next tick
                Bukkit.getScheduler().scheduleSyncDelayedTask(plugin) {
                    BmPlayerLeaveGameIntent.leave(player)
                }
            }
            player.gameMode = GameMode.SURVIVAL
            player.exhaustion = 0f
            player.foodLevel = 100000 // just a big number
            player.isFlying = false
            player.inventory.clear()
            for ((i, stack) in game.settings.initialItems.withIndex()) {
                if (i < player.inventory.size)
                    player.inventory.setItem(i, stack?.clone())
            }
            removePotionEffects(player)

            // Add tag for customisation
            player.addScoreboardTag("bm_player")
        }

        fun bombStrength(game: Game, player: Player): Int {
            var strength = 1
            for (stack in player.inventory.contents) {
                if (stack != null && stack.type == game.settings.powerItem) {
                    strength += stack.amount
                }
            }
            return strength.coerceAtLeast(1)
        }

        @JvmStatic
        fun setupLoginWatcher() {
            Bukkit.getPluginManager().registerEvents(object: Listener {
                @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
                fun onPlayerLogin(e: PlayerJoinEvent) {
                    val player = e.player
                    if (player.isDead)
                        return // cannot reset a dead player (client side glitches out)
                    val save = tempDataFile(player)
                    if (save.exists())
                        reset(player)
                }
                @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
                fun onPlayerRespawn(e: PlayerRespawnEvent) {
                    val player = e.player
                    val save = tempDataFile(player)
                    if (save.exists())
                        e.respawnLocation = reset(player)
                }
            }, plugin)
        }

        private fun reset(player: Player): Location {
            // Give player StuffBack
            val file =  tempDataFile(player)
            val dataFile = YamlConfiguration.loadConfiguration(file)

            ((dataFile["gamemode"] as? String?)?.let { try {
                GameMode.valueOf(it.uppercase())
            } catch (e: IllegalArgumentException) {
                null
            }} ?: GameMode.SURVIVAL).let { player.gameMode = it }
            (dataFile["health-scale"] as? Number? ?: 20).let { player.healthScale = it.toDouble() }
            val maxHealth = player.getAttribute(Attribute.GENERIC_MAX_HEALTH)!!
            maxHealth.baseValue = (dataFile["health-max"] as? Number? ?: 20.0).toDouble()
            maxHealth.modifiers.forEach { maxHealth.removeModifier(it) }
            (dataFile["health"] as? Number? ?: 20).let { player.health =
                // Health may be greater than max health if modifiers were applied before joining
                it.toDouble().coerceAtMost(maxHealth.value)
            }
            (dataFile["food-level"] as? Number? ?: 20).let { player.foodLevel = it.toInt() }
            (dataFile["inventory"] as? List<Any?> ?: emptyList())
                    .map {
                        if (it is ItemStack) it else null
                    }.toTypedArray()
                    .let { player.inventory.contents = it }
            (dataFile["is-flying"] as? Boolean? ?: false).let { player.isFlying = it }
            val location = (dataFile["location"] as? Location?
                    ?: Bukkit.getServer().worlds.first().spawnLocation)
            player.teleport(location)

            player.removeScoreboardTag("bm_player")

            try {
                player.addAttachment(plugin, "group.bomberman", false)
            } catch (e: Exception) {
                plugin.logger.log(Level.WARNING, "Unable to remove permissions", e)
            }

            file.delete()

            removePotionEffects(player)
            return location
        }

        private fun removePotionEffects(player: Player) {
            Bukkit.getScheduler().scheduleSyncDelayedTask(plugin) {
                player.fireTicks = 0
                for (effect in player.activePotionEffects) {
                    player.removePotionEffect(effect.type)
                }
            }
        }

        private fun tempDataFile(player: Player): File {
            return File(plugin.settings.tempPlayerData(), "${player.name}.yml")
        }
    }

    private var immunity = false

    /**
     * Removes the player from the game and removes any hooks to this player. Treats the player like they disconnected
     * from the server.
     */
    private fun resetStuffAndUnregister() {
        // remove items in the direct vicinity (prevents player dropping items at spawn)
        player.world.getNearbyEntities(player.location, 1.0, 2.0, 1.0).stream()
                .filter { it is ItemStack }
                .forEach { it.remove() }

        reset(player)
        HandlerList.unregisterAll(this)
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    fun onPlayerJoinGame(e: BmPlayerJoinGameIntent) { // Cannot join two games at once
        if (e.player === player) {
            e.cancelFor(Text.JOIN_ALREADY_JOINED
                    .with("game", e.game)
                    .with("player", player)
                    .format())
        }
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun onCount(e: BmTimerCountedEvent) {
        if (e.game != game)
            return
        if (e.count > 0) {
            Text.GAME_COUNT
                    .with("time", e.count)
                    .with("game", game)
                    .sendTo(player)
        }
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun onRunStarted(e: BmRunStartedIntent) {
        if (e.game != game)
            return
        Text.GAME_STARTED
                .with("game", game)
                .sendTo(player)
        e.setHandled()
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOWEST)
    fun onPlayerRegen(e: EntityRegainHealthEvent) {
        if (e.entity != player)
            return
        if (e.regainReason == EntityRegainHealthEvent.RegainReason.MAGIC) {
            e.amount = 1.0
        } else {
            e.isCancelled = true
        }
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    fun onPlayerBreakBlockWithWrongTool(e: PlayerInteractEvent) {
        if (e.player !== player)
            return
        if (e.action != Action.LEFT_CLICK_BLOCK || !e.hasBlock()) {
            // Only care about block breaking events
            return
        }

        // Players can only break block if they have used the correct tool
        // This simulates Adventure Mode with the CanDestroy nbt tags
        // Note - we do not put the player in Adventure because then they cannot place blocks
        val key = e.clickedBlock?.blockData?.material?.key?.toString()
        if (e.item != null && key != null) {
            val list = BukkitAdapter.adapt(e.item).nbtData?.getList("CanDestroy", StringTag::class.java)
            if (!list.isNullOrEmpty()) {
                if (list
                    .map { it.value }
                    .any { key.equals(it, ignoreCase = true) }
                ) {
                    // allow breakage
                    return
                }
            }
        }

        // Cancel break event
        e.isCancelled = true
        e.setUseInteractedBlock(Event.Result.DENY)
        e.setUseItemInHand(Event.Result.DENY)
        // apply mining fatigue so player doesn't see block breaking
        e.player.addPotionEffect(PotionEffect(PotionEffectType.SLOW_DIGGING, 20, 1))
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    fun onPlayerPlaceBlock(e: BlockPlaceEvent) {
        if (e.player != player)
            return

        // if the block has the CanPlaceOn tag, restrict it to those types
        val key = e.blockAgainst.blockData.material.key.toString()
        val tagList = BukkitAdapter.adapt(e.itemInHand).nbtData?.getList("CanPlaceOn", StringTag::class.java)
        if (!tagList.isNullOrEmpty()) {
            if (tagList
                    .map { it.value }
                    .none { key.equals(it, ignoreCase = true) }
            ) {
                // stop breakage
                e.isCancelled = true
            }
        }
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
    fun onPlayerPlaceTNT(e: BlockPlaceEvent) {
        if (e.player !== player) return
        val b = e.block
        // create a bomb when placing tnt
        if (b.type == game.settings.bombItem) {
            if (!Bomb.spawnBomb(game, player, b)) {
                e.isCancelled = true
            }
        }
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    fun onPlayerMoved(e: PlayerMoveEvent) {
        if (e.player !== player)
            return
        val bmEvent = BmPlayerMovedEvent(game, player, e.from, e.to ?: e.from)
        Bukkit.getPluginManager().callEvent(bmEvent)
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun onExplosion(e: BmExplosionEvent) {
        if (e.game != game)
            return
        // TODO duplicate code: both GamePlayer and Explosion do touching checks
        if (Explosion.isTouching(player, e.igniting.map(Explosion.BlockPlan::block).toSet())) {
            BmPlayerHitIntent.hit(player, e.cause)
        }
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun onPlayerHit(e: BmPlayerHitIntent) {
        if (e.player !== player)
            return
        BmPlayerHurtIntent.run(game, player, e.cause)
        e.setHandled()
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    fun onPlayerHurtWithImmunity(e: BmPlayerHurtIntent) {
        if (e.player !== player)
            return
        if (immunity)
            e.isCancelled = true
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun onPlayerDamaged(e: BmPlayerHurtIntent) {
        if (e.player !== player)
            return
        if (player.health > 1) {
            player.damage(1.0)
            immunity = true
            player.fireTicks = game.settings.immunityTicks
            Bukkit.getScheduler().scheduleSyncDelayedTask(plugin, {
                immunity = false
                player.fireTicks = 0
                // Call the player move event to recheck damage required
                Bukkit.getPluginManager().callEvent(
                        BmPlayerMovedEvent(game, player, player.location, player.location))
            }, game.settings.immunityTicks.toLong()) // 22 is slightly longer than 20 ticks a bomb is active for
        } else {
            BmPlayerKilledIntent.kill(game, player, e.attacker)
        }
        e.setHandled()
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun onPlayerKilledInGame(e: BmPlayerKilledIntent) {
        if (e.player !== player) return
        player.health = 0.0
        BmPlayerLeaveGameIntent.leave(player)
        e.setHandled()
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun onPlayerWon(e: BmPlayerWonEvent) {
        if (e.player !== player) return
        Text.PLAYER_WON
                .with("player", player)
                .sendTo(player)
        // Let player walk around like a boss until the game stops
        immunity = true
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGH)
    fun onPlayerLeaveGameEvent(e: BmPlayerLeaveGameIntent) {
        if (e.player !== player)
            return

        // Remove items near the player (stops players duplicating items at spawn)
        player.world.getNearbyEntities(player.location, 2.0, 3.0, 2.0)
                .filterIsInstance<Item>()
                .forEach{ it.remove() }

        // Give player their stuff back
        if (player.isDead) {
            // Attempting reset player health when dead causes very strange bugs.
            // Just let the loginWatcher handle it
            HandlerList.unregisterAll(this)
        } else {
            resetStuffAndUnregister()
        }
        e.setHandled(game)
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
    fun onGameStopped(e: BmRunStoppedIntent) {
        if (e.game != game)
            return
        BmPlayerLeaveGameIntent.leave(player)
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
    fun onGameTerminated(e: BmGameTerminatedIntent) {
        if (e.game != game)
            return
        // Note this and gameStop will not double process leave event
        // since the leave intent unregisters the handler
        BmPlayerLeaveGameIntent.leave(player)
    }

    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    fun onPlayerLogout(e: PlayerQuitEvent) {
        if (e.player == player) {
            BmPlayerLeaveGameIntent.leave(player)
        }
    }
}